### PR TITLE
Allow custom python versions and environments

### LINKED ISSUE #448: Impossible to use version number in Ubuntu 16.04
- Puppet: 5.5.2
- Ruby: ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]
- Distribution: Ubuntu 16.04.2 LTS
- Module version: v2.2.2

I want to make puppet install python3 on a host and use pip3 provider

##  My manifest:
```puppet
 class { 'python':
    ensure => 'present',
    version => '3',
    pip     => 'latest',
  }
  python::pip { $package_id:
    ensure       => 'latest',
    pkgname      => $package_name,
    pip_provider => 'pip3',
  }
```
##  puppet run output error
```
Error: Could not update: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 3' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package 3
Error: /Stage[main]/Python::Install/Package[python]/ensure: change from purged to latest failed: Could not update: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 3' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package 3
```

If I try to use 
```puppet
version => 'python3'
``` 
I get:
```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, version needs to be within[system, pypy, 3, 3.3, 2.7]
``` 

Is it a bug or I miss something here?

I am new to puppet but I am just wondering when I will meet the condition here: https://github.com/voxpupuli/puppet-python/blob/f274e51ffdf53c5f17d468053d27977c586ae8ef/manifests/install.pp#L236

### PR DESCRIPTION
**Pull request by jradmacher**
#### Pull Request (PR) description
This pull request allows to install all python versions and create **pyenv** or **virtualenv** environments
with those versions. 
This also adds the missing _venv_ module in newer Debian based distributions and addresses the deprecation of _pyenv_ in 3.6 with an expected removal of it in Python 3.8.

#### This Pull Request (PR) fixes the following issues
 Fixes #448
 Fixes the first part of #430 "pip3 missing in the virtualenv ?"

### CODE REVIEW

**Comment by bastelfreak** [comment]
Hi @jradmacher, could you please squash the commits down a bit? Otherwise this looks good!

**Comment by bastelfreak** [comment]
Please also have a look at the used email address in the commits. It isn't associated with your github account.

**Comment by jradmacher** [comment]
Squashed to 4 commits. I don't want create fewer commits, as each commit introduces a non trivial behavior change.

Could you rerun the TravisCI build? It failed with an unrelated VM error.

File: manifests/init.pp [hunk 1]
```diff
@@ -62,8 +62,9 @@
   }
 
   $allowed_versions = concat(['system', 'pypy'], $valid_versions)
```
Comment (bastelfreak) [inline manifests/init.pp]: this line is now obsolete?

Comment (jradmacher) [reply manifests/init.pp]: Thanks, I missed this line.

**Comment by jradmacher** [comment]
Updates:
- rebased to current master
Our internal test runs were successful as were your Travis CI checks.

**Comment by bastelfreak** [comment]
thanks!

**Pull request merged**